### PR TITLE
[Dashboard] Move gcs health check to a separate thread to avoid crashing due to excessive CPU usage.

### LINKED ISSUE #17964: Dashboard crashing due to GCS RPC errors
Context:

@TanjaBayer running Ray Serve on Kubernetes (need to verify exact version). The dashboard crashes but their Serve endpoints are still reachable, so direct actor communication is working at least. The dashboard logs show the following:

```
2021-08-19 05:05:47,915 INFO node_head.py:291 -- Received error entry for 10.233.124.245, repr=<ray.serve.controller.ServeController object at 0x7fafb10438d0> 2979
2021-08-19 05:05:47,918 ERROR head.py:67 -- Got AioRpcError when checking GCS is alive, seq=108.
Traceback (most recent call last):
  File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/new_dashboard/head.py", line 59, in _gcs_check_alive
    request, timeout=2)
  File "/home/ray/anaconda3/lib/python3.7/site-packages/grpc/aio/_call.py", line 286, in __await__
    self._cython_call._status)
grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
        status = StatusCode.DEADLINE_EXCEEDED
        details = "Deadline Exceeded"
        debug_error_string = "{"created":"@1629374736.314417484","description":"Deadline Exceeded","file":"src/core/ext/filters/deadline/deadline_filter.cc","file_line":81,"grpc_status":4}"
>
2021-08-19 05:05:47,918 ERROR head.py:74 -- Dashboard suicide, the GCS RPC error count 11 > 10
```

and there are repeated log messages like this:
```
2021-08-19 05:05:47,915 INFO node_head.py:291 -- Received error entry for 10.233.124.245, repr=<ray.serve.controller.ServeController object at 0x7fafb10438d0> 2979
```

### PR DESCRIPTION
**Pull request by simon-mo**
Serve users reported cases of dashboard crashing with error message
```
2021-08-30 13:02:18,725	ERROR head.py:72 -- Got AioRpcError when checking GCS is alive, seq=77.
Traceback (most recent call last):
  File "..../venv/lib/python3.7/site-packages/ray/new_dashboard/head.py", line 64, in _gcs_check_alive
    request, timeout=2)
  File "..../venv/lib/python3.7/site-packages/grpc/aio/_call.py", line 286, in __await__
    self._cython_call._status)
grpc.aio._call.AioRpcError: <AioRpcError of RPC that terminated with:
	status = StatusCode.DEADLINE_EXCEEDED
	details = "Deadline Exceeded"
	debug_error_string = "{"created":"@1630321317.983901658","description":"Deadline Exceeded","file":"src/core/ext/filters/deadline/deadline_filter.cc","file_line":81,"grpc_status":4}"
>
2021-08-30 13:02:18,725	ERROR head.py:80 -- Dashboard exiting because it received too many GCS RPC errors count: 11, threshold is 10.
```

Furthere investigation of the log found that the dashboard head was overloaded by other modules. In this case, the dashboard head module was processing error entry non-stop (repeated every 30ms) 
```
2021-08-30 13:02:18,723	INFO node_head.py:292 -- Received error entry for 172.16.13.182, repr=<ray.serve.controller.ServeController object at 0x7f0a04598790> 237491
```

The hypothesis is that the gRPC timeout comes from dashboard head being CPU contended, especially in asyncio mode, the event loop is blocked by node_head. Therefore the health check timeout. I furthere investigated that gRPC asyncio does have this issue when the asyncio loop is blocked by CPU intensive work. Even when the health check response returned before the timeout, if loop is busy, DEADLINE_EXCEEDED will still be raised. See grpc reproduction here: https://gist.github.com/simon-mo/788b02fbf7d19eaf7c19106cdb1ec470

The fix, therefore, is to move GCS health check to a separate thread to make sure the health check does not suffer than loop blocking issue by other dasboard head modules. Alternatively, I can create a synchrounous gRPC client within the health checker thread but that's roughly the same amount of work.

Closes #17964

### CODE REVIEW

File: src/ray/protobuf/gcs_service.proto [hunk 1]
```diff
@@ -190,12 +190,10 @@ message ReportHeartbeatReply {
 }
 
 message CheckAliveRequest {
-  int32 seq = 1;
```
Comment (simon-mo) [inline src/ray/protobuf/gcs_service.proto:193]: this field has not been used in any substantial way. the seq number that's echo-ed back is never consumed by client.

**Review by edoakes — COMMENTED** [review]
@simon-mo looks reasonable, is it possible to test that this fixes the issue before merging?

**Comment by fyrestone** [comment]
Are you sure the excessive CPU usage is because of gcs health checking? In our practice (the same interval as the community), the gcs health checking does not cost many CPUs. And it makes the dashboard complex by mixing threading and asyncio.

**Comment by simon-mo** [comment]
@fyrestone please check the description for the reasoning behind this change. 

@edoakes I tested it manually with the following repro:
```
import ray
import time
import logging

ray.init(log_to_driver=False)

@ray.remote(num_cpus=0)
class A:
    def __init__(self):
        logger = logging.getLogger("ray")
        while True:
            logger.error("hiiiiiiiiiii")
            time.sleep(0.001)

actors = [A.remote() for _ in range(8)]

while True:
    time.sleep(0.01)v
```
(dashboard crashed with DEADLINE_EXCEEDED before the fix and ok after)

**Review by wuisawesome — APPROVED** [review]

**Review by edoakes — APPROVED** [review]

**Comment by fyrestone** [comment]
> @fyrestone please check the description for the reasoning behind this change.
> 
> @edoakes I tested it manually with the following repro:
> 
> ```
> import ray
> import time
> import logging
> 
> ray.init(log_to_driver=False)
> 
> @ray.remote(num_cpus=0)
> class A:
>     def __init__(self):
>         logger = logging.getLogger("ray")
>         while True:
>             logger.error("hiiiiiiiiiii")
>             time.sleep(0.001)
> 
> actors = [A.remote() for _ in range(8)]
> 
> while True:
>     time.sleep(0.01)v
> ```
> 
> (dashboard crashed with DEADLINE_EXCEEDED before the fix and ok after)

Is the error message pubsubed to the dashboard? We have disabled the _update_log_info() and _update_error_info: https://github.com/ray-project/ray/blob/master/dashboard/modules/node/node_head.py#L301 because pubsub logs and errors costs too many CPUs in large cluster. I think this PR does not fix the root cause of excessive CPU usage.

**Comment by simon-mo** [comment]
I agree. This doesn't fix the root cause, but it does make sure the health check behaves correctly and not being impacted by any other components of the dashboard.

**Comment by fyrestone** [comment]
> I agree. This doesn't fix the root cause, but it does make sure the health check behaves correctly and not being impacted by any other components of the dashboard.

I agree. This PR makes the health check more stable.

**Pull request merged**